File: aulas/20151210/01-Serial/EchoServer_Serial.java
import java.net.*;
import java.nio.*;
import java.nio.channels.*;
import java.util.concurrent.*;

public final class EchoServer_Serial
{
   private static int sessionCount = 0;

   public static void main(String[] args) throws Exception
   {
      ServerSocketChannel server =
         ServerSocketChannel.open().bind(new InetSocketAddress(8888));

      System.out.println(":: EchoServer running ::");

      for (;;) {
         try {
            SocketChannel session = server.accept();
            int sessionId = ++sessionCount;
            
            attend(sessionId, session);

         } catch (Exception exc) {
            System.err.println(":: EchoServer error: " + exc.getMessage());
            exc.printStackTrace();
            System.exit(1);
         }
      }
   }

   private static void attend(final int sessionId, final SocketChannel session) throws Exception
   {
      System.out.printf("++ [S%d:T%d] connection from %s\n",
         sessionId,
         Thread.currentThread().getId(),
         session.getRemoteAddress()
      );
         
      int nb;
      final ByteBuffer buffer = ByteBuffer.allocate(8);
         
      for (;;) {
         try {
            nb = session.read(buffer);
                
            if (nb == -1) {
               System.err.printf("++ [S%d:T%d] end of session error\n",
                  sessionId,
                  Thread.currentThread().getId()
               );
               try { session.close(); } catch (Exception e) {}
               return;
            }
                   
            System.err.printf("++ [S%d:T%d] processing %d byte(s)\n",
               sessionId,
               Thread.currentThread().getId(),
               nb
            );

            try { Thread.sleep(500); } catch (Exception e) {}

            buffer.flip();
            try { session.write(buffer); } catch (Exception e) {}

            buffer.clear();
               
         } catch (Exception exc) {
            System.err.printf("++ [S%d:T%d] session error: %s\n",
               sessionId,
               Thread.currentThread().getId(),
               exc.getMessage());
            exc.printStackTrace();
            return;
         }
      }
   }
}

File: aulas/20151210/01-Serial/EchoServer_Serial.cs
using System;
using System.Net.Sockets;
using System.Threading;
using System.Text;
using System.Net;

public static class EchoServer_Serial
{
    private static int SessionCount = 0;

    public static void Main()
    {
        TcpListener server = new TcpListener(IPAddress.Any, 8888);
        server.Start();

        Console.WriteLine(":: EchoServer running ::");

        for (;;) { 
            TcpClient session = server.AcceptTcpClient();
            int sessionId = ++SessionCount;

            Attend(sessionId, session);
        }
    }

    private static void Attend(int sessionId, TcpClient session)
    {
        Console.WriteLine("++ [S{0}:T{1}] connection from {2}",
            sessionId,
            Thread.CurrentThread.ManagedThreadId,
            session.Client.RemoteEndPoint
        );
        NetworkStream stream = session.GetStream();

        int nb;
        byte[] buffer = new byte[8];

        for (;;) {
            nb = stream.Read(buffer, 0, buffer.Length);

            if (nb == 0) {
                Console.WriteLine("++ [S{0}:T{1}] end of session",
                    sessionId,
                    Thread.CurrentThread.ManagedThreadId
                );
                session.Close();
                return;
            }

            Console.WriteLine("++ [S{0}:T{1}] processing {2} byte(s)",
                sessionId,
                Thread.CurrentThread.ManagedThreadId,
                nb
            );

            Thread.Sleep(500);

            stream.Write(buffer, 0, nb);
        }
    }
}

